### PR TITLE
[forward-port] add auto fill popup: Copy cells and Fill series

### PR DESCRIPTION
**Pull request by bayramcicek**
AutoFill contextmenu has two options:

- Copy Cells
- Fill Series

- requires 2 upstream patches: 
  - https://gerrit.libreoffice.org/c/core/+/167726 (MERGED to `master`, `23.05`, `24.04`) - 5ab9b63e6624947a6076a01f9bab0d7a925e0e40
  - https://gerrit.libreoffice.org/c/core/+/168086 - f7345a0d4b24210d27326e5d9acd5ba92d468f1d

* Resolves: # 
* Target version: master
* see also: (`23.05`) https://github.com/CollaboraOnline/online/pull/9039

### Summary

- Added UI
- close AutoFill contextmenu when
  - switch tabs by clicking on a different tab button
  - switch tabs by `Ctrl+Alt+PgUp/Down`
- Added functionalities for "Copy Cells" and "Fill Series".
- close AutoFill contextmenu when contextmenu opens from sheet tabs

### TODO

- [ ] ...

### Checklist

- [x] Code is properly formatted
- [x] All commits have Change-Id
- [ ] I have run tests with `make check`
- [x] I have issued `make run` and manually verified that everything looks okay
- [x] Documentation (manuals or wiki) has been updated or is not required

### CODE REVIEW

**Comment by bayramcicek** [comment]
@eszkadev : Hi. I'd like to have your review on this PR. 

If there are small changes needed I can do it immediately. Otherwise; If there are no blockers, let's merge, and I'd like to implement your reviews with a follow-up patch since I work on another accessibility bug now. 

Thanks in advance :)

File: browser/src/control/Control.AutoFillOptions.ts [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+ * Copyright the Collabora Online contributors.
+ *
+ * SPDX-License-Identifier: MPL-2.0
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+/*
+ * Control.AutoFillOptions - class for auto fill popup
```
Comment (eszkadev) [inline browser/src/control/Control.AutoFillOptions.ts:12]: best to write how to found it so someone new will know

File: browser/src/map/handler/Map.Keyboard.js [hunk 2]
```diff
@@ -770,6 +770,10 @@ L.Map.Keyboard = L.Handler.extend({
 						else {
 							partToSelect = currentSelectedPart != 0 ? currentSelectedPart - 1 : parts - 1;
 						}
+
+						if (this._map._docLayer.isAutoFillPopupOpen)
+							this._map.fire('closeautofillpopup');
```
Comment (eszkadev) [inline browser/src/map/handler/Map.Keyboard.js]: please just fire an event and do the "if ( ... ) " in the handler

File: browser/src/control/Control.Tabs.js [hunk 3]
```diff
@@ -324,6 +324,9 @@ L.Control.Tabs = L.Control.extend({
 		if (part !== this._map._docLayer._selectedPart) {
 			this._setPartIndex(part);
 		}
+
+		if (this._map._docLayer.isAutoFillPopupOpen)
```
Comment (eszkadev) [inline browser/src/control/Control.Tabs.js]: please just fire an event and do the "if ( ... ) " in the handler

File: browser/src/control/Control.JSDialog.js [hunk 4]
```diff
@@ -602,7 +602,8 @@ L.Control.JSDialog = L.Control.extend({
 			var dialog = this.dialogs[autoFilterDialogId];
 
 			// Check if the current dialog has the isAutofilter property set to true
-			if (dialog.isAutofilter) {
+			// and if it is not an autoFillPopup
+			if (dialog.isAutofilter && (autoFilterDialogId !== 'autoFillPopup')) {
```
Comment (eszkadev) [inline browser/src/control/Control.JSDialog.js]: why is that needed? what it does? it doesn't look right

File: browser/src/control/Control.AutoFillOptions.ts [hunk 5]
```diff
@@ -76,20 +75,20 @@ class AutoFillOptions extends L.Control.AutoCompletePopup {
 		data.posx = framePos.x;
 		data.posy = framePos.y;
 		this.sendJSON(data);
+
+		this.map._docLayer.isAutoFillPopupOpen = true;
```
Comment (eszkadev) [inline browser/src/control/Control.AutoFillOptions.ts]: Please do not move implementation details outside of the class. We can store that information here - or not store but just check.

Please add `isOpen()` method to the base class, which can be used instead maybe?

File: browser/src/control/Control.AutoFillOptions.ts [hunk 6]
```diff
@@ -51,8 +51,8 @@ class AutoFillOptions extends L.Control.AutoCompletePopup {
 	}
 
 	getPopupEntries(): any[] {
-		const entries: any[] = [{ text: '', columns: [{ text: 'Copy cells' }], row: '' },
-		{ text: '', columns: [{ text: 'Fill series' }], row: '' }];
+		const entries: any[] = [{ text: 'copy', columns: [{ text: 'Copy cells' }], row: '0' },
```
Comment (eszkadev) [inline browser/src/control/Control.AutoFillOptions.ts]: Labels should be alwasy translated:
`'abc' -> _('abc')`

File: browser/src/control/Control.AutoFillOptions.ts [hunk 7]
```diff
@@ -113,7 +113,10 @@ class AutoFillOptions extends L.Control.AutoCompletePopup {
 			}
 
 		} else if (eventType === 'keydown') {
-			// TODO: handle arrowDown key
+			if (object.key !== 'Tab' && object.key !== 'Shift') {
+				this.map.focus();
+				return true;
+			}
```
Comment (eszkadev) [inline browser/src/control/Control.AutoFillOptions.ts:120]: it shouldn't be much different that in the base class - can we call the base class method?

File: browser/src/control/Control.AutoFillOptions.ts [hunk 8]
```diff
@@ -51,8 +51,10 @@ class AutoFillOptions extends L.Control.AutoCompletePopup {
 	}
 
 	getPopupEntries(): any[] {
-		const entries: any[] = [{ text: 'copy', columns: [{ text: 'Copy cells' }], row: '0' },
-		{ text: 'fill', columns: [{ text: 'Fill series' }], row: '1' }];
+		const entries: any[] = [
+			{ text: 'copy', columns: [{ text: 'Copy cells' }], row: '0' },
```
Comment (eszkadev) [inline browser/src/control/Control.AutoFillOptions.ts:55]: translation

File: browser/src/control/Control.JSDialog.js [hunk 9]
```diff
@@ -600,10 +600,17 @@ L.Control.JSDialog = L.Control.extend({
 		for (var i = 0; i < dialogKeys.length; i++) {
 			var autoFilterDialogId = dialogKeys[i];
 			var dialog = this.dialogs[autoFilterDialogId];
+			var isAutoFillFromOnMouseUp = this.map._docLayer.isAutoFillFromOnMouseUp;
```
Comment (eszkadev) [inline browser/src/control/Control.JSDialog.js:603]: I don't think we should have `isAutofilter == true` for that popup....
How we set this up? is that needed? maybe just don't use it for autofill?

File: browser/src/control/Control.AutoFillOptions.ts (hunk 1 above)
Comment (bayramcicek) [reply browser/src/control/Control.AutoFillOptions.ts:12]: done

File: browser/src/map/handler/Map.Keyboard.js (hunk 2 above)
Comment (bayramcicek) [reply browser/src/map/handler/Map.Keyboard.js]: this is removed in the new commits: no need

File: browser/src/control/Control.Tabs.js (hunk 3 above)
Comment (bayramcicek) [reply browser/src/control/Control.Tabs.js]: this is removed in the new commits: no need x 2

File: browser/src/control/Control.AutoFillOptions.ts (hunk 5 above)
Comment (bayramcicek) [reply browser/src/control/Control.AutoFillOptions.ts]: this is removed in the new commits: no need x 4

File: browser/src/control/Control.AutoFillOptions.ts (hunk 6 above)
Comment (bayramcicek) [reply browser/src/control/Control.AutoFillOptions.ts]: I only changed `text: 'Copy cells'` and `text: 'Fill series'` since they are only visible to user.
(`text: 'copy'` and `text: 'fill'` stay the same as before)

File: browser/src/control/Control.AutoFillOptions.ts [hunk 10]
```diff
@@ -95,11 +95,21 @@ class AutoFillOptions extends L.Control.AutoCompletePopup {
...
 			if (index == 0) {
-				// Copy cells
-				this.map.sendUnoCommand('.uno:FillSeries?FillStep:string=0');
+				/*
+					- Copy cells
+					undo needed here, "Copy cells" should operate on the original content,
+					not the new content of the cells.
+				*/
+				this.map.sendUnoCommand('.uno:Undo');
+				this.map.sendUnoCommand('.uno:AutoFill?Copy:bool=true');
```
[the remark this answers is not in the document]
Comment (bayramcicek) [reply browser/src/control/Control.AutoFillOptions.ts:106]: `undo` calls are removed on the latest commit

File: browser/src/control/Control.AutoFillOptions.ts (hunk 7 above)
Comment (bayramcicek) [reply browser/src/control/Control.AutoFillOptions.ts]: I moved `if (object.key !== 'Tab' && object.key !== 'Shift')` to the base class.

File: browser/src/control/Control.AutoFillOptions.ts (hunk 8 above)
Comment (bayramcicek) [reply browser/src/control/Control.AutoFillOptions.ts]: done

File: browser/src/control/Control.JSDialog.js (hunk 9 above)
Comment (bayramcicek) [reply browser/src/control/Control.JSDialog.js:603]: I think this is very useful since when sheet tabs are changed, it automatically closes the tabs (including mine - autoFillPopup) so we don't need to fire `close` for every situation that we need to close popup. 

This code block also runs after when we use `autoFill` with `AutoFill marker box`. Since I open the popup before the cells are filled with data, this for loop closes `autoFill` popup immediately after it opens, therefore we can't see the popup. 

That's why I want to prevent that `this.close(autoFilterDialogId, true)` call only if we use `autoFill` with `AutoFill marker box`. (which triggers with `onMouseUp` in `browser/src/canvas/sections/AutoFillMarkerSection.ts`)

Comment (eszkadev) [reply browser/src/control/Control.JSDialog.js:603]: I agree with the effect, but I'm concerned about code, if it is not misleading :)

My question is: why    `isAutofilter` is set for auto fill popup which is not `autofilter` ?

Comment (bayramcicek) [reply browser/src/control/Control.JSDialog.js:603]: To be honest. I actually don't know why :) I think mentionPopup and formulaAutoComplete popup is also `true` for ` isAutofilter`.

Comment (eszkadev) [reply browser/src/control/Control.JSDialog.js:603]: `instance.isAutofilter = instance.isDocumentAreaPopup && this.map._docLayer.isCalc();`  in Control.JSDialog.js

Could you put your check for `id` there? so it will NOT be `isAutofilter == true` ?

File: browser/src/layer/tile/CanvasTileLayer.js [hunk 11]
```diff
@@ -3204,6 +3204,12 @@ L.CanvasTileLayer = L.Layer.extend({
 			var topLeftPixels = this._twipsToCorePixels(topLeftTwips);
 			var offsetPixels = this._twipsToCorePixels(offset);
 			this._cellAutoFillAreaPixels = L.LOUtil.createRectangle(topLeftPixels.x, topLeftPixels.y, offsetPixels.x, offsetPixels.y);
+
+			// if comes from onMouseUp event of AutoFillMarkerSection.ts, open AutoFill popup with location data
+			if (this._map._docLayer.isAutoFillFromOnMouseUp) {
+				// TODO: don't open AutoFill popup if empty cells are selected
```
[the remark this answers is not in the document]
Comment (bayramcicek) [reply browser/src/layer/tile/CanvasTileLayer.js:3210]: will be done in a follow up asap.

File: browser/src/control/Control.JSDialog.js (hunk 9 above)
Comment (bayramcicek) [reply browser/src/control/Control.JSDialog.js]: Thanks, I have missed that part. 
Done with -> https://github.com/CollaboraOnline/online/pull/9135/commits/1bdaba888f3964d8c74bc2fc1572d3545109d7ea

File: browser/src/control/Control.JSDialog.js [hunk 12]
```diff
@@ -606,6 +606,16 @@ L.Control.JSDialog = L.Control.extend({
 				// Call this.close(key, true) for the current dialog
 				this.close(autoFilterDialogId, true);
 			}
+
+			if (autoFilterDialogId === 'autoFillPopup') {
```
Comment (eszkadev) [inline browser/src/control/Control.JSDialog.js:610]: this is no longer needed as we fixed problem with `isAutofilter`

Comment (eszkadev) [reply browser/src/control/Control.JSDialog.js:610]: or am I wrong?

Comment (bayramcicek) [reply browser/src/control/Control.JSDialog.js:610]: This is needed. I don't want to put `closePopup` manually everywhere when there is a need to close popup on tab change. Here is easy, just don't close the autoFilter popup only if we drag with autoFilterMarker. otherwise close it. :)

Comment (eszkadev) [reply browser/src/control/Control.JSDialog.js:610]: ah, it's needed... then please add `instance.isAutoFillPopup` property and detect it here. (We want to avoid using the hardcoded id in many places).
Also change name of the function to `closePopupsOnTabChange` so it will tell the truth.

File: browser/src/control/Control.JSDialog.js [hunk 13]
```diff
@@ -639,7 +649,7 @@ L.Control.JSDialog = L.Control.extend({
 		instance.isDocumentAreaPopup = instance.popupParent === '_POPOVER_' && instance.posx !== undefined && instance.posy !== undefined;
 		instance.isPopup = instance.isModalPopUp || instance.isDocumentAreaPopup || instance.isSnackbar;
 		instance.containerParent = instance.isDocumentAreaPopup ? document.getElementById('document-container'): document.body;
-		instance.isAutofilter = instance.isDocumentAreaPopup && this.map._docLayer.isCalc();
+		instance.isAutofilter = instance.isDocumentAreaPopup && this.map._docLayer.isCalc() && instance.id !== 'autoFillPopup';
```
Comment (eszkadev) [inline browser/src/control/Control.JSDialog.js:652]: here add `instance.isAutoFillPopup`, and also use it in `isAutofilter` to not duplicate detection

Comment (bayramcicek) [reply browser/src/control/Control.JSDialog.js]: done

File: browser/src/control/Control.JSDialog.js (hunk 12 above)
Comment (bayramcicek) [reply browser/src/control/Control.JSDialog.js]: Thanks ! 
Done

**Review by eszkadev — APPROVED** [review]
code looks ok now

**Comment by bayramcicek** [comment]
> autofill-2024-05-24_16.17.05.webm
> 
>     1. click on a corner of a cell
> 
>     2. click anywhere
>        Result: autofill popup appears
> 
> 
> Is that expected?

No. I have to prevent `isAutoFillFromOnMouseUp = true` if there is no dragging. `onMouseUp` called when we do that. then when we select a random cell it thinks it comes from autoFill.

**Comment by bayramcicek** [comment]
> autofill-2024-05-24_16.17.05.webm
> 
>     1. click on a corner of a cell
> 
>     2. click anywhere
>        Result: autofill popup appears
> 
> 
> Is that expected?

Done

**Review by eszkadev — APPROVED** [review]
Looks ok as a first step

**Comment by eszkadev** [comment]
I think detection based on booelan when we used "handle" is fragile.
I managed to not show the auto fill popup if I moved cursor very fast.

Can we later trigger it from the core? so we don't need that boolean anymore? What do you think?

**Comment by bayramcicek** [comment]
> I think detection based on booelan when we used "handle" is fragile. I managed to not show the auto fill popup if I moved cursor very fast.
> 
> Can we later trigger it from the core? so we don't need that boolean anymore? What do you think?

it is because of the issue on onMouseUp event in the `browser/src/canvas/sections/AutoFillMarkerSection.ts`:

>> I see that mouseButtonDown/Up called only if cursor is over the AutoFillMarker box. The crosshair can also seen outside of the AutoFillMarker box:
>> Therefore, as far as I can see/understand;
>> - if mouse cursor is very close to the AutoFillMarker box, the crosshair become visible and AutoFill operation executes.
>> - BUT mouseButtonDown/Up is not called there.


> This is interesting. I think we can make the size of the autofill section larger. Hopefully something like that will solve the issue.

**Comment by bayramcicek** [comment]
> I think detection based on booelan when we used "handle" is fragile. I managed to not show the auto fill popup if I moved cursor very fast.
> 
> Can we later trigger it from the core? so we don't need that boolean anymore? What do you think?

I'm doing some changes in core - which seems will makes us get rid of that boolean...

**Comment by bayramcicek** [comment]
> I think detection based on booelan when we used "handle" is fragile. I managed to not show the auto fill popup if I moved cursor very fast.
> 
> Can we later trigger it from the core? so we don't need that boolean anymore? What do you think?

@eszkadev : I added a core change : https://gerrit.libreoffice.org/c/core/+/168086 which helped us to NOT depend on `OnMouseUp` and `isAutoFillFromOnMouseUp`. 

(Tests are failed here because the patch is requires core change)

Ready for review. Thanks :)

File: browser/src/layer/tile/CanvasTileLayer.js [hunk 14]
```diff
@@ -1439,6 +1439,9 @@ L.CanvasTileLayer = L.Layer.extend({
 		else if (textMsg.startsWith('cellautofillarea:')) {
 			this._onCellAutoFillAreaMsg(textMsg);
 		}
+		else if (textMsg.startsWith('cellautofillpopuponmouseup:')) {
```
Comment (eszkadev) [inline browser/src/layer/tile/CanvasTileLayer.js:1442]: more generic name please (no on mouse up in the name)

**Comment by eszkadev** [comment]
Thank you, now it looks more bulletproof :)

File: browser/src/control/Control.AutoFillOptions.ts [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+			this.closeMentionPopup({ typingMention: false } as CloseMessageEvent);
+		} else if (eventType === 'select' || eventType === 'activate') {
+			this.map.fire('closeautofillpopup');
+
+			if (index == 0) {
+				// Copy cells
+				this.map.sendUnoCommand('.uno:AutoFill?Copy:bool=true');
+			} else if (index == 1) {
+				// Fill series
+				this.map.sendUnoCommand('.uno:AutoFill?Copy:bool=false');
```
Comment (eszkadev) [inline browser/src/control/Control.AutoFillOptions.ts:106]: These actions will only do "uno command" calls? We don't need anything additional there?
Now I see we could reuse ... existing context menu  for it then if we don't need any custom content in that popup.

But this PR makes some good cleanup in TextInput.js

File: browser/src/layer/tile/CanvasTileLayer.js (hunk 14 above)
Comment (bayramcicek) [reply browser/src/layer/tile/CanvasTileLayer.js]: DONE

File: browser/src/control/Control.AutoFillOptions.ts (hunk 15 above)
Comment (eszkadev) [reply browser/src/control/Control.AutoFillOptions.ts:106]: Please see example CALLBACK and it's JSON when right click on cell:
```
contextmenu: {
    "menu": [
        {
            "text": "~Cut",
            "type": "command",
            "command": ".uno:Cut",
            "enabled": "true"
        },
```

could you check if we can just reuse LOK_CALLBACK_CONTEXT_MENU  which already shows a menu and it can be shown in the correct place?

Then we will avoid custom callback.

Comment (eszkadev) [reply browser/src/control/Control.AutoFillOptions.ts:106]: You can just send `command` field without `text`, in the online we can get correct translatable string for it.

context menu is implemented in `browser/src/control/Control.ContextMenu.js`

File: browser/src/layer/tile/CanvasTileLayer.js [hunk 16]
```diff
@@ -3216,6 +3216,7 @@ L.CanvasTileLayer = L.Layer.extend({
 			var topLeftPixels = this._twipsToCorePixels(topLeftTwips);
 			var offsetPixels = this._twipsToCorePixels(offset);
 			this._cellAutoFillAreaPixels = L.LOUtil.createRectangle(topLeftPixels.x, topLeftPixels.y, offsetPixels.x, offsetPixels.y);
+			this._map._docLayer.autoFillLocation = { data: topLeftPixels };
```
Comment (eszkadev) [inline browser/src/layer/tile/CanvasTileLayer.js:3219]: not used

File: browser/src/layer/tile/CalcTileLayer.js [hunk 17]
```diff
@@ -88,6 +88,7 @@ L.CalcTileLayer = L.CanvasTileLayer.extend({
 		app.sectionContainer.addSection(new app.definitions.AutoFillMarkerSection());
 
 		this.insertMode = false;
+		this.autoFillLocation = {};
```
Comment (eszkadev) [inline browser/src/layer/tile/CalcTileLayer.js:91]: not used

**Review by eszkadev — COMMENTED** [review]
Thanks, now I'm happy with the solution. Thanks for that many updates :)

File: browser/src/layer/tile/CalcTileLayer.js (hunk 17 above)
Comment (bayramcicek) [reply browser/src/layer/tile/CalcTileLayer.js]: DONE

File: browser/src/layer/tile/CanvasTileLayer.js (hunk 16 above)
Comment (bayramcicek) [reply browser/src/layer/tile/CanvasTileLayer.js]: DONE

**Comment by bayramcicek** [comment]
> Thanks, now I'm happy with the solution. Thanks for that many updates :)

Thanks for the reviews and support! :)

**Review by eszkadev — APPROVED** [review]
